### PR TITLE
JDK-8209968: Fix off by one error in image scale calculation

### PR DESCRIPTION
**Pull request by hjohn**
See issue #164.

Contains additional test for case when dimensions are computed with one of them set to 0, just to confirm that they're working correctly.

### CODE REVIEW

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import com.sun.javafx.iio.common.ImageTools;
+import static org.junit.Assert.*;
+import org.junit.Test;
+
+import java.util.Random;
+
+public class ImageToolsTest {
+    @Test
+    public void testIfComputeDimensionsReturnsCorrectValuesWhenAspectRatioIsPreserved() {
+        for (int i = 2; i < 1000; i += 2) {
```
Comment (eugener) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: Lot of "magic" numbers. Having constants with some explanatory names would be desirable.

Comment (eugener) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: Is there any need to have long loops ( 1000s of iterations ) for  this kind of test?

**Comment by kevinrushforth** [comment]
In addition to answering Eugene's comments, you have a white-space error in the newly-added test file (see the Travis build log).

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java (hunk 1 above)
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: I don't like this test case very much at all.  I'm having trouble thinking of one that would a) cover all edge cases, and b) does not use the same algorithm in the test itself.  Covering all edge cases is tough since you take 4 int parameters, making this a huge problem space, not just the usual of  covering a couple of branches.

If a test is sufficient that shows the old code breaking and the new code working, then I'll be happy to rewrite it.  As it is, I'm unsure how to make it cover all cases without a couple of loops.

Comment (eugener) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: I have no problem with loops at all. But do we need thousands of iterations?

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import org.junit.Test;
+
+import java.util.Random;
+
+public class ImageToolsTest {
+    private static final int RANDOM_SEED = 1;  // A random seed
+    private static final int MIN_SIZE = 1;  // Smallest dimension to generate
+    private static final int MAX_SIZE = 1000;  // Largest dimension to generate
+    private static final int RANGE = MAX_SIZE - MIN_SIZE;  // Generate dimension range
+    private static final int RANDOM_COUNT = 10000;  // Number of random source and target dimensions to generate and verify
```
[the remark this answers is not in the document]
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: It was running for 0.12s, I reduced it to 1000 loops.

I managed to come up with asserts that verify if it is closely matching the intended aspect, instead of just some boundary tests.  This catches the known failure cases (which were also caught already by the `assertTrue(msg, x == tw || y == th)` assert) without having to hard code any specific cases.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java (hunk 1 above)
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: I've simplified the entire test.

Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: I've reduced the number of iterations, and added some additional asserts that accurately assert the expected values for the dimension that didn't reach it's maximum size.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/iio/common/ImageTools.java [hunk 3]
```diff
@@ -667,8 +667,8 @@ private static void computeUpdatedPixels(int sourceOffset,
                     finalHeight = (int) ((float) sourceHeight * finalWidth / sourceWidth);
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/iio/common/ImageTools.java]: I think the above two blocks should also round instead of truncate. Otherwise the smaller dimension will be calculated differently depending on whether the requested size is 0 or is a larger-than-needed number.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+public class ImageToolsTest {
+    private static final int RANDOM_SEED = 1;  // A random seed
+    private static final int MIN_SIZE = 1;  // Smallest dimension to generate
+    private static final int MAX_SIZE = 1000;  // Largest dimension to generate
+    private static final int RANGE = MAX_SIZE - MIN_SIZE;  // Generate dimension range
+    private static final int RANDOM_COUNT = 1000;  // Number of random source and target dimensions to generate and verify
+
+    @Test
+    public void testIfComputeDimensionsReturnsValuesInCorrectRangeWhenAspectRatioIsPreserved() {
+        Random rnd = new Random(RANDOM_SEED);
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: I still want to see some tests with explicit values, including a few with one dimension being 0. You might create a utility method that takes the 4 values, which can then be called with a few carefully chosen values and then also called in the loop with random values. Especially in light of the comment I made on the fix itself.

As an example, the following will produce different results because the first will (after your fix) be rounded:

```
sx = 98, sy = 97, mx = 40, my= 50 -- result: x = 40, y = 40 (with your fix)
sx = 98, sy = 97, mx = 40, my= 0 -- result: x = 40, y = 39 (with or without your fix)
```

File: modules/javafx.graphics/src/main/java/com/sun/javafx/iio/common/ImageTools.java (hunk 3 above)
Comment (hjohn) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/iio/common/ImageTools.java]: Alright, I rounded this as well, and covered these in the unit test.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java (hunk 4 above)
Comment (hjohn) [reply modules/javafx.graphics/src/test/java/test/com/sun/javafx/iio/common/ImageToolsTest.java]: I update the test with your suggestions, and included sufficient cases to cover all branches of the entire function.

**Comment by kevinrushforth** [comment]
Since this is close to being ready to go in, can you send a review request email to openjfx-dev@openjdk.java.net referencing the JBS bug and this pull request? That way others can see it and comment if they want.
